Remove repo_root requirement, make build more hermetic, cacheable

cromwellian opened the pull request:
Repo_root no longer needed

General description:
1) Add filegroup to bring node_modules/ into bazel
2) Add all_srcs and other filegroups for TS sources needed
3) Map all files needed by sigh command into runfiles/ via deps attribute
4) runfiles becomes the new PWD during sigh execution, and the symlink farm created by bazel looks like the repository root
5) Removes all execution_requiremens except local (regretably) due to some issues with node_modules/ dir

overall, this makes the build more hermetic since it doesn't mutate the source repo.

Ultimately, this should also allow remote_cache and remote_execution with bazel, to speed up the CI, and it's closer to the way g3 builds work.

See https://nicolovaligi.com/faster-bazel-remote-caching-benchmark.html for example of a 50x speedup.

piotrswigon commented:
I'll defer to Cameron, I wouldn't be able to review this properly :(

cromwellian commented on build_defs/build_defs.bzl:
```diff
@@ -17,14 +18,14 @@ def _run_schema2pkg(name, src, out, language_name, language_flag, file_extension
+              "--outdir $(dirname $PWD/{OUT}) " +
               "--outfile $(basename {OUT}) " +
-              "{SRC}",
+              "$PWD/{SRC}",
```
The reason why this is prefixed by $PWD (absolutize the src path) is because $PWD is the execroot.  The problem with bazel and tools/sigh is that it always changes the CWD to the repo root, but SRC is relative to your workspace root (e.g. particles/PipeApps/Foo.arcs). But, if you have multiple workspaces, then files in an external workspace are symlinked into an 'external/' subdirectory under exec_root.  Thus, SRC = external/particles/PipeApps/Foo.arcs, and this will fail. 

By absolutizing the file, schema2kotlin Manifest.load() will work with relative imports, regardless of the CWD.

csilvestrini replied:
Thanks for the explanation. Could you add a short comment to the code so that others who read the code later can see it too?

cromwellian commented on src/javaharness/java/arcs/demo/particles/CaptureKotlinEntity.kt:
```diff
@@ -0,0 +1,39 @@
+
+import arcs.api.*
+import arcs.demo.services.ClipboardService
+
```
This is just a placeholder automatic conversion of CaptureEntity to Kotlin, it does not use the generated schema2kotlin entity yet, I merely added this to test transitive Kotlin library builds w/arcs_kt_schema

csilvestrini replied:
Could you add that as a comment to the file (with a TODO to delete it)

csilvestrini commented on WORKSPACE:
```diff
@@ -6,3 +8,19 @@ emsdk_repo()
+
+load("@io_bazel_rules_kotlin//kotlin:kotlin.bzl", "kotlin_repositories", "kt_register_toolchains")
+kotlin_repositories() # if you want the default. Otherwise see custom kotlinc distribution below
+kt_register_toolchains() # to use the default toolchain, otherwise see toolchains below
```
These comments don't seem necessary. Just delete them?

csilvestrini commented on build_defs/build_defs.bzl:
```diff
@@ -17,14 +18,14 @@ def _run_schema2pkg(name, src, out, language_name, language_flag, file_extension
         srcs = [src],
         outs = [out],
-        deps = ["//src/tools:schema2pkg_srcs"],
+        deps = ["@arcs//src/tools:schema2pkg_srcs", "@arcs//tools"],
```
How come you've added the `@arcs` prefix here? Shouldn't these targets work happily without that?

cromwellian replied:
There's some weird stuff going on with multi-workspace builds I'm trying to sort out. 

Because we're using build macros, not rules, when you import a build macro from the Arcs workspace into the arcs_javaharness workspace,  it's deps will be stuff like '//src' or '//build_defs', but inside of the javaharness workspace, those don't exist. 

This is the problem with using macros, they don't know about workspaces, they're just python functions. 

This is causing so much trouble, I'm thinking about eliminating the multi-workspace, and changing all BUILD files to be based off of the root workspace.

csilvestrini commented on build_defs/run_in_repo.bzl:
```diff
@@ -9,9 +9,9 @@ EXECUTION_REQUIREMENTS_TAGS = [
+def _absolute_path(path):
     """Converts path relative to the repo root into an absolute file path."""
-    return repo_root + "/" + path
+    return path
```
Just delete this function

alxrsngrtn replied:
Agreed

csilvestrini commented on src/wasm/kotlin/BUILD:
```diff
@@ -1,7 +1,15 @@
+
+kt_jvm_library(
     name = "arcs",
+    srcs = glob(["src/arcs/Entity.kt"]),
```
I feel like this file doesn't belong inside the `wasm` directory if it is generally useful outside of wasm code. There should probably be a general Kotlin folder somewhere, and the wasm-specific stuff can depend on it. Doesn't have to be in this PR, but a TODO at the least would be good.

alxrsngrtn replied:
Ray had the idea of adding a `src/sdk` directory (i.e. this could go under `src/sdk/Kotlin`). Is it time to make the move?

cromwellian replied:
Let's defer this and discuss in a doc, since there's now a src_kt dir as well, and the copybara script is getting complicated, so we need to rethink repo layout for non-ts

csilvestrini commented on build_defs/run_in_repo.bzl:
```diff
@@ -21,40 +21,28 @@ def _write_shell_script(ctx, run_script):
     """
 
-    # Extract the repo root directory from .bazelrc
-    if "repo_root" not in ctx.var:
```
If repo_root isn't needed anymore, could you please also remove it from tools/setup?

csilvestrini commented on build_defs/build_defs.bzl:
```diff
@@ -17,14 +18,14 @@ def _run_schema2pkg(name, src, out, language_name, language_flag, file_extension
         # TODO: generated header guard should contain whole workspace-relative
         # path to file.
-        cmd = "./tools/sigh schema2pkg " +
+        cmd = "$(location @arcs//tools) schema2pkg " +
```
This doesn't seem to work. When I try testing locally I get:

```
==================== Test output for //src/wasm/cpp/tests:wasm-cpp-test:
/usr/local/google/home/csilvestrini/.cache/bazel/_bazel_csilvestrini/3b285f724e0b71687487035dc2909458/execroot/arcs/bazel-out/k8-fastbuild/bin/src/wasm/cpp/tests/wasm-cpp-test.sh.runfiles/arcs/src/wasm/cpp/tests/wasm-cpp-test.sh: line 3:
./tools/sigh: No such file or directory
================================================================================
```

csilvestrini replied:
It also seems to be failing on Travis (with a different error though):

```
ENOENT: no such file or directory, open '/home/travis/build/PolymerLabs/arcs/home/travis/.cache/bazel/_bazel_travis/504270edcd9a48d6d285f3ee471e3abc/execroot/arcs/src/wasm/cpp/tests/schemas.arcs'
Post-parse processing error caused by './in-memory-1773926.manifest' line 1.
Error importing '/home/travis/build/PolymerLabs/arcs//home/travis/.cache/bazel/_bazel_travis/504270edcd9a48d6d285f3ee471e3abc/execroot/arcs/src/wasm/cpp/tests/schemas.arcs'
  import '/home/travis/build/PolymerLabs/arcs//home/travis/.cache/bazel/_bazel_travis/504270edcd9a48d6d285f3ee471e3abc/execroot/arcs/src/wasm/cpp/tests/schemas.arcs'
  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
No schemas found in '/home/travis/.cache/bazel/_bazel_travis/504270edcd9a48d6d285f3ee471e3abc/execroot/arcs/src/wasm/cpp/tests/schemas.arcs'
```

cromwellian replied:
Yeah, it's weird, works on Kotlin if you build from the src/javaharness directory, but not from outside. It appears related to nested workspaces, I'm trying to fix it now. I think if we get rid of the nested WORKSPACE, it would be simpler, but then non-nested WORKSPACE leads to to nastier paths everywhere.

alxrsngrtn replied:
Shouldn't it be: `$(location @arcs//tools)/sigh`?
Does `@arcs/tools` point to sigh specifically or the tools directory?

alxrsngrtn replied:
Just read the tools BUILD file. 

Does it make sense to have it be `${location @arcs//tools):sigh`?

cromwellian replied:
its fixed now, the only restriction is "local" execution due to Docker privileges for creating symlinks.

csilvestrini commented on build_defs/run_in_repo_template.sh:
```diff
@@ -1,5 +1,3 @@
 set -e
-# cd to repo root first before running command.
-cd "{repo_root}"
 {cmd}
```
It doesn't look like this file is needed at all anymore. Can you delete it, and instead of running `ctx.actions.expand_template` just run the command instead (using `ctx.actions.run` or `ctx.actions.run_shell`).

cromwellian replied:
Seems like arcs_ts_test still wants a shell script returned in DefaultInfo provider (e.g. executable=script_file), so may need to keep it?

alxrsngrtn commented on WORKSPACE:
```diff
@@ -6,3 +8,19 @@ emsdk_repo()
+rules_kotlin_sha = "245d0bc1511048aaf82afd0fa8a83e8c3b5afdff0ae4fbcae25e03bb2c6f1a1a"
+http_archive(
+    name = "io_bazel_rules_kotlin",
+    urls = ["https://github.com/cgruber/rules_kotlin/archive/%s.zip" % rules_kotlin_version],
```
Should we try to unify the kotlin-native vs kotlin dependency?

cromwellian replied:
We could fork the official repo like cgruber did, but I think we need to think through the toolchain / auto_wasm stuff before we go there.

alxrsngrtn commented on build_defs/build_defs.bzl:
```diff
@@ -42,14 +43,20 @@ def arcs_cc_schema(name, src, out = None):
+    kt_jvm_library(
+        name = name,
+        srcs = [out],
+        deps = ["@arcs//src/wasm/kotlin:arcs"]
```
I see how this ties in to the particle / Wasm disentanglement work that you said I should do. I'll refer back here for reference.

alxrsngrtn commented on build_defs/build_defs.bzl:
```diff
@@ -42,14 +43,20 @@ def arcs_cc_schema(name, src, out = None):
     """Generates a Kotlin file for the given .arcs schema file."""
     _run_schema2pkg(
-        name = name,
+        name = name + "_genrule",
```
Does this require that we change existing BUILD files that make use of this macro?

alxrsngrtn replied:
For example, `particles/Native/Wasm/BUILD`

alxrsngrtn replied:
Also `particles/Tutorial/Kotlin/BUILD`

alxrsngrtn commented on tools/BUILD:
```diff
@@ -0,0 +1,9 @@
+
+filegroup(
+    name = "tools",
+    srcs = (["sigh"]),
```
Ah. This answers my question from above.

cromwellian commented:
I have it working now, but only with spawn_strategy=local, which still, in a way, relies on being able to see the original 'repo' instead of the sandbox. I have it partially working inside the sandbox, which will be more ideal, since if it works inside the sandbox, it'll work with remote execution and workers, which will allow build-farming to speed up the build.

So not ready yet.

cromwellian commented:
Ok Cameron (or Jason), PTAL. 

I simplified and removed some of the hacks. 

Basically, the gist is to make build rules dep on node_modules and src/ directories. This causes symlinks to be created in runfiles, which means the build macros can execute sigh mapped into the sandbox via local paths instead of relying on global absolute references to the repo, or blaze-out. 

This CL will enable us to turn on remote_cache and remote_build more easily, to try and speed up the bazel tests.

csilvestrini commented on build_defs/build_defs.bzl:
```diff
@@ -112,7 +118,13 @@ def arcs_ts_test(name, src, deps):
+            "//tools:sigh_bin",
+            "//tools:tools_srcs",
+            "//:node_modules",
+            "//:tsconfig"
```
There's a lot of duplicated deps here just needed to run sigh correctly. Adding a separate rule in tools/ to run sigh subcommands might be a good idea. e.g. 

```
sigh_command(
    name = name,
    command = "test --bazel --file {SRC}",
    ...
)
```

csilvestrini commented on javaharness/java/arcs/demo/particles/CaptureEntityFactory.java:
```diff
@@ -0,0 +1,31 @@
+package arcs.demo.particles;
```
This file doesn't seem relevant?

csilvestrini commented on shells/pipes-shell/BUILD:
```diff
@@ -6,12 +6,24 @@ run_in_repo(
     visibility = ["//visibility:public"],
     deps = [
-        "//tools:sigh_webpack",
+        "//tools:sigh_bin", 
```
I don't think this is right. The old version used to run sigh webpack, before running deploy.sh. The Arcs runtime needs to be built before you can deploy the shell. If you change runtime code and then run this build rule again, it won't rebuild anything.

I think you need to revert this to depend on sigh_webpack

csilvestrini commented on tools/BUILD:
```diff
@@ -1,13 +1,30 @@
+    deps = ["//tools:sigh_bin",
+            "//tools",
+            "@npm//:node_modules",
+            "//:tsconfig"],
```
Can you run buildifier on all BUILD and .bzl files?

csilvestrini commented on src/BUILD:
```diff
@@ -1,11 +0,0 @@
-filegroup(
-    name = "core_srcs",
```
You can't delete this, it's needed by sigh_webpack

cromwellian replied:
sigh_command uses //:all_srcs which includes core_srcs, which basically means bazel will create a symlink farm from repo_root into runfiles, which is similar to how run_in_repo root semantics works, only with more isolation.

csilvestrini commented on shells/pipes-shell/BUILD:
```diff
@@ -6,12 +6,24 @@ run_in_repo(
+
+filegroup(
+    name = "web_deploy",
+    srcs = glob(["web/deploy/**"]),
```
Doesn't seem like there's a need for this. There's already a `srcs = glob(["**"])` line above, so these files are already being pulled in

cromwellian replied:
$location expansion doesn't work on a glob that has more than one file, so you need a target that has just one file, however you're right in that it is being double included, so the target above should exclude deploy.sh

csilvestrini commented on build_defs/sigh.bzl:
```diff
@@ -0,0 +1,34 @@
+        srcs = srcs,
+        outs = outs,
+        # TODO: generated header guard should contain whole workspace-relative
+        # path to file.
```
This TODO is only relevant for schema2pkg (for C++). You can just delete it from here.
